--- cgiwrap.c.orig	2009-05-18 21:33:58.740663288 +0200
+++ cgiwrap.c	2009-05-18 21:34:44.687797891 +0200
@@ -199,6 +199,8 @@
 			StringEndsWith(scriptPath, ".php") ||
 			StringEndsWith(scriptPath, ".php3") ||
 			StringEndsWith(scriptPath, ".php4") ||
+			StringEndsWith(scriptPath, ".php5") ||
+			StringEndsWith(scriptPath, ".phpc") ||
 			StringEndsWith(scriptPath, ".phtml") )
 		{
 			Context.interpreted_script = 1;
