--- fetch.c.orig	2010-04-21 16:22:19.904682203 +0200
+++ fetch.c	2010-04-21 16:24:39.937167976 +0200
@@ -30,10 +30,13 @@
 	char *pathInfoString;
 	char *queryString;
 	char *userStr;
+	char *phprc;
+	char *phpdir;
 
 	DEBUG_Msg("\n");
 
 	userStr = (char *) 0;
+	phpdir = "/etc/php/users.d/";
 	//pathInfoString = getenv("PATH_INFO");
 	pathInfoString = getenv("PATH_TRANSLATED");
 	if ( pathInfoString )  /* use PATH_INFO */
@@ -44,6 +47,11 @@
 			DEBUG_Msg("Trying to extract user from PATH_TRANSLATED.");
 
 			userStr = GetPathComponents(1, pathInfoString);
+			phprc = (char *)malloc((strlen(phpdir) + strlen(userStr) + 1) *sizeof(char));
+			strcpy (phprc , phpdir);
+			strcat (phprc , userStr);
+			setenv("PHPRC", phprc, 1);
+			DEBUG_Str("Setting PHPRC to ", phprc);
 		}
 		else
 		{
